firebase-firestore/src/main/java/com/google/firebase/firestore/model/ObjectValue.java

```diff
@@ -153,16 +170,18 @@ public void delete(FieldPath path) {
    */
   public void set(FieldPath path, Value value) {
     hardAssert(!path.isEmpty(), "Cannot set field for empty path on ObjectValue");
-    setOverlay(path, value);
+    synchronized (overlayMap) {
+      setOverlay(path, value);
+    }
   }
 
   public void setAll(Map<FieldPath, Value> data) {
-    for (Map.Entry<FieldPath, Value> entry : data.entrySet()) {
-      FieldPath path = entry.getKey();
-      if (entry.getValue() == null) {
-        delete(path);
-      } else {
-        set(path, entry.getValue());
+    synchronized (overlayMap) {
+      hardAssert(
```

Comment:
Was it intentional to include the `hardAssert` within the synchronized block?

Reply:
Good point. I will move it out!